charts/redpanda/service.loadbalancer.go

```diff
@@ -0,0 +1,175 @@
+//
+// +gotohelm:filename=service.loadbalancer.go.tpl
+package redpanda
+
+import (
+	"fmt"
+
+	"github.com/redpanda-data/helm-charts/pkg/gotohelm/helmette"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+)
+
+func LoadBalancerServices(dot *helmette.Dot) []*corev1.Service {
+	values := helmette.Unwrap[Values](dot.Values)
+
+	// This is technically a divergence from previous behavior but this matches
+	// the NodePort's check and is more reasonable.
+	if !values.External.Enabled || !values.External.Service.Enabled {
+		return nil
+	}
+
+	if values.External.Type != corev1.ServiceTypeLoadBalancer {
+		return nil
+	}
+
+	externalDNS := ptr.Deref(values.External.ExternalDNS, Enableable{})
+
+	labels := FullLabels(dot)
+	// This typo is intentionally being preserved for backwards compat
+	// https://github.com/redpanda-data/helm-charts/blob/2baa77b99a71a993e639a7138deaf4543727c8a1/charts/redpanda/templates/service.loadbalancer.yaml#L33
+	labels["repdanda.com/type"] = "loadbalancer"
+
+	selector := StatefulSetPodLabelsSelector(dot)
+
+	services := []*corev1.Service{}
+	//   {{- range $replicaIndex := untilStep 0 ($values.statefulset.replicas|int) 1 }}
+	replicas := values.Statefulset.Replicas // TODO fix me once the transpiler is fixed.
+	for i := 0; i < replicas; i++ {
+		//     {{- $podName := printf "%s-%d" (include "redpanda.fullname" $root) $replicaIndex }}
+		podname := fmt.Sprintf("%s-%d", Fullname(dot), i)
+
+		// NB: A range loop is used here as its the most terse way to handle
+		// nil maps in gotohelm.
+		annotations := map[string]string{}
+		for k, v := range values.External.Annotations {
+			annotations[k] = v
+		}
+
+		if externalDNS.Enabled {
+			prefix := podname
+			if len(values.External.Addresses) > i {
+				prefix = values.External.Addresses[i]
+			}
+
+			address := fmt.Sprintf("%s.%s", prefix, helmette.Tpl(*values.External.Domain, dot))
+
+			annotations["external-dns.alpha.kubernetes.io/hostname"] = address
+		}
+
+		// NB: A range loop is used here as its the most terse way to handle
+		// nil maps in gotohelm.
+		podSelector := map[string]string{}
+		for k, v := range selector {
+			podSelector[k] = v
+		}
+
+		podSelector["statefulset.kubernetes.io/pod-name"] = podname
+
+		ports := []corev1.ServicePort{}
+
+		for name, listener := range values.Listeners.Admin.External {
+			if !ptr.Deref(listener.Enabled, values.External.Enabled) {
+				continue
+			}
+
+			fallbackPorts := append(listener.AdvertisedPorts, values.Listeners.Admin.Port)
+
+			ports = append(ports, corev1.ServicePort{
+				Name:     fmt.Sprintf("admin-%s", name),
+				Protocol: corev1.ProtocolTCP,
+				// TODO this diverges from all the other cases that use
+				// listener.Port. Is this a bug in the original template?
+				TargetPort: intstr.FromInt32(values.Listeners.Admin.Port),
+				Port:       ptr.Deref(listener.NodePort, fallbackPorts[0]),
```

Comment:
I wonder how `values.Listeners.Admin.External` could be deterministic with ranging and always returning the same port. 

Would there be a case where running helm template in a loop without changing values input would generate different loadbalancer manifest?

Reply:
Ah, I'm actually going to punt on this change for now. The nodeport service works the same way and templates' `range`s are actually deterministic.